pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,359 @@
+		}
+		glog.V(5).Infof("PersistentVolume[%s] is now available\n", volume.Name)
+		nextPhase = api.VolumeAvailable
+
+	// available volumes await a claim
+	case api.VolumeAvailable:
+		if volume.Spec.ClaimRef != nil {
+			_, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+			if err == nil {
+				// change of phase will trigger an update event with the newly bound volume
+				glog.V(5).Infof("PersistentVolume[%s] is now bound\n", volume.Name)
+				nextPhase = api.VolumeBound
+			} else {
+				if errors.IsNotFound(err) {
+					nextPhase = api.VolumeReleased
+				}
+			}
+		}
+	//bound volumes require verification of their bound claims
+	case api.VolumeBound:
+		if volume.Spec.ClaimRef == nil {
+			return fmt.Errorf("PersistentVolume[%s] expected to be bound but found nil claimRef: %+v", volume)
+		} else {
+			claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+			if err == nil {
+				// bound and active.  Build claim status as needed.
+				if claim.Status.VolumeRef == nil {
+					syncClaimStatus(binderClient, volume, claim)
+				}
+			} else {
+				if errors.IsNotFound(err) {
+					nextPhase = api.VolumeReleased
+				} else {
+					return err
+				}
+			}
+		}
+	// released volumes require recycling
+	case api.VolumeReleased:
+		if volume.Spec.ClaimRef == nil {
+			return fmt.Errorf("PersistentVolume[%s] expected to be bound but found nil claimRef: %+v", volume)
+		} else {
+			// TODO: implement Recycle method on plugins
+		}
+	}
+
+	if currentPhase != nextPhase {
+		volume.Status.Phase = nextPhase
+		volume, err := binderClient.UpdatePersistentVolumeStatus(volume)
+		if err != nil {
+			// Rollback to previous phase
+			volume.Status.Phase = currentPhase
+		}
+		volumeIndex.Update(volume)
+	}
+
+	return nil
+}
+
+func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, claim *api.PersistentVolumeClaim) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolumeClaim[%s]\n", claim.Name)
+
+	// claims can be in one of the following states:
+	//
+	// ClaimPending -- default value -- not bound to a claim.  A volume that matches the claim may not exist.
+	// ClaimBound -- bound to a volume.  claim.Status.VolumeRef != nil
+	currentPhase := claim.Status.Phase
+	nextPhase := currentPhase
+
+	switch currentPhase {
+
+	// pending claims await a matching volume
+	case api.ClaimPending:
+		volume, err := volumeIndex.FindBestMatchForClaim(claim)
+		if err != nil {
+			return err
+		}
+		if volume == nil {
+			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
+		}
+
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim.
+		// triggers update of the volume in this controller, which builds claim status
+		volume.Spec.ClaimRef = claimRef
```

Comment:
As per the binding sub-resource discussion, I sense that we have this backwards.  PersVolume.Status should host a ClaimRef and PersVolumeClaim.Spec should host a VolumeName (not ref, I think).